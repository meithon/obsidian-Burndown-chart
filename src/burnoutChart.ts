

/**
 
```dataviewjs
const s = require(app.vault.adapter.basePath + "/.scripts/index.cjs");
s.burndown(dv, app, this.container, "2024-07-21")
```

**/
export async function burndown(dv, app, container, targetEndDateString: string) {
  const tasks = await getTasks(dv)
  console.log(tasks)
  const data = burnoutDataFromTasks(tasks, new Date(targetEndDateString))

  renderBurndownChart(
    app,
    container,
    data,
  )
}


function formatDatetoMMDD(date: Date) {
  return new Intl.DateTimeFormat('en-US', { month: 'short', day: '2-digit' }).format(date);
}




export function burnoutDataFromTasks(tasks: Task[], targetEndDate: Date) {
  const { oldestTask, newestTask } = getStartAndEnd(tasks, targetEndDate)

  const startDate = oldestTask.completion ?? new Date()
  const remainPriod = getPriodDate(startDate, newestTask.completion ?? targetEndDate)
  const planePriod = getPriodDate(startDate, targetEndDate)

  return {
    ramianTasksPerDay: getRemainTasksPerDay(tasks, remainPriod),
    plannedTasksPerDay: getPlandedTasksPerDay(tasks, planePriod),
    labels: remainPriod.map(formatDatetoMMDD),
    tasksAmount: tasks.length,
  };
}
type BurnoutData = ReturnType<typeof burnoutDataFromTasks>

async function getTasks(dv) {
  const queryResult = await dv.tryQuery(
    'TASK WHERE contains(file.path, this.file.path)'
  ) as QueryResult
  const tasks = (queryResult.values).map(t => {
    if (!t.completion) {
      const { completion, ...rest } = t
      return {
        ...rest,
        completion: undefined,
      }
    }
    const { year, month, day } = t.completion.c
    return {
      ...t,
      completion: new Date(`${year}-${month}-${day}`)
    }
  })
  return tasks
}

const dayMilisec = 1000 * 60 * 60 * 24

function getStartAndEnd(tasks: Task[], targetEndDate: Date) {
  const sorted = tasks.sort((a, b) => {
    if (!a.completion) return 0
    if (!b.completion) return 0

    return a.completion.getTime() > b.completion.getTime() ? 1 : -1
  })
  const newestTask = sorted[sorted.length - 1]
  const oldestTask = sorted[0]

  return {
    newestTask,
    oldestTask
  }
}

function getPriodDate(startDate: Date, endDate: Date): Date[] {
  const days: Date[] = []
  const diffDays = Math.floor((endDate.getTime() - startDate.getTime()) / dayMilisec) + 1
  for (let i = 0; i < diffDays; i++) {
    const tmpDate = new Date(startDate)
    tmpDate.setDate(tmpDate.getDate() + i)
    days.push(tmpDate)
  }
  return days
}





function getRemainTasksPerDay(tasks: Task[], days: Date[]): number[] {
  const completedTasks = days.map(day => {
    const completedTasks = tasks.filter(task => {
      if (task.completion === undefined) return false
      return task.completion.getTime() <= day.getTime()
    })
    return completedTasks.length
  })
  const remainTasks = completedTasks.map((i) => tasks.length - i)
  return remainTasks
}

function getPlandedTasksPerDay(tasks: Task[], days: Date[]): number[] {
  const taskPerDay = tasks.length / days.length
  return days.map((_, i) => {
    return Math.round((tasks.length - taskPerDay * (i + 1)) * 100) / 100
  })
}



interface Dataview {
  tryQuery: (query: string) => Promise<QueryResult>
}



function renderBurndownChart(
  app,
  container,
  data: BurnoutData
) {
  const {
    ramianTasksPerDay: remain,
    plannedTasksPerDay: planned,
    labels,
    tasksAmount
  } = data

  const option = {
    width: 700,
    height: 250,
    backgroundColor: 'transparent',
    grid: {
      bottom: -1,
      containLabel: true
    },
    title: {
      text: ''
    },
    toolbox: {
      left: 250,
      feature: {
        saveAsImage: {}
      }
    },
    tooltip: { trigger: 'axis' },
    legend: { left: 70 },
    xAxis: {
      type: 'category',
      name: 'Day',
      axisTick: { length: 0.06 },
      axisLabel: { rotate: 0, interval: 0, textStyle: { fontSize: "11", color: "#fff" }, height: 100 },
      data: ["start", ...labels]
    },
    yAxis: {
      name: "Story points",
      axisLabel: { textStyle: { fontSize: "11", color: "#fff" } }
    },
    series: [
      {
        type: "line",
        color: '#ff598a',
        name: "Planned",
        data: [
          tasksAmount,
          ...planned,
        ]
      },
      {
        type: "line",
        color: '#59ff6d',
        name: "Actual",
        data: [
          tasksAmount,
          ...remain
        ]
      },

    ]

  }
  app.plugins.plugins['obsidian-echarts'].render(
    option,
    container
  )
}

// ```dataviewjs
// const tasks = await	dv.tryQuery('TASK WHERE contains(file.path, this.file.path)')
// console.log(JSON.stringify(tasks, null, 2))
// ```


// type QueryResult = typeof sampleTask;

interface DateTime {
  c: {
    day: number,
    hour: number,
    millisecond: number,
    minute: number,
    month: number,
    second: number,
    year: number,
  }
}

type QueryResult = {
  values: {
    completion?: DateTime,
    completed: boolean
  }[]
}
// type Task = QueryResult['values']

export interface Task {
  completion?: Date,
  completed: boolean

}



export const _test = {
  getPriod: getStartAndEnd,
  getPriodDate,
  getRemainTasksPerDay,
  getPlandedTasksPerDay,
}
