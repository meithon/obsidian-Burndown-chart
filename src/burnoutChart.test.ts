import { expect, test, describe } from "bun:test";
import { _test, burnoutDataFromTasks, Task } from "./burnoutChart";

const date = (date: string): Date => new Date(date)

const task = (dateString?: string): Task => {
  const completion = dateString ? date(dateString) : undefined

  return {
    completed: true,
    completion,
  }

}

describe("4 tasks", () => {
  test("burnoutDataFromTasks", () => {
    const tasks = [
      task("2023-01-01"),
      task("2023-01-02"),
      task("2023-01-03"),
      task("2023-01-04"),
    ]

    const res = burnoutDataFromTasks(tasks, date("2023-01-04"))
    expect(res).toEqual(
      {
        "ramainTasksPerDay": [3, 2, 1, 0],
        "labels": [
          "Jan 01",
          "Jan 02",
          "Jan 03",
          "Jan 04",
        ],
        "plannedTasksPerDay": [3, 2, 1, 0],
        tasksAmount: 4,
      }
    )
  });
  test("burnoutDataFromTasks - 1 task", () => {
    const tasks = [
      "2023-01-01",
    ].map((data) => task(data))

    const res = burnoutDataFromTasks(tasks, date("2023-01-04"))
    expect(res).toEqual(
      {
        "ramainTasksPerDay": [0],
        "labels": [
          "Jan 01",
        ],
        "plannedTasksPerDay": [0],
        tasksAmount: 1,
      }
    )
  });
  test("burnoutDataFromTasks - 2 task", () => {
    const tasks = [
      "2023-01-01",
      "2023-01-02",
    ].map((data) => task(data))

    const res = burnoutDataFromTasks(tasks, date("2023-01-04"))
    expect(res).toEqual(
      {
        ramainTasksPerDay: [1, 0],
        labels: [
          "Jan 01",
          "Jan 02",
        ],
        plannedTasksPerDay: [1, 0],
        tasksAmount: 2,
      }
    )
  });

  test("getPriod", () => {
    const oldestTask = task("2023-01-01")
    const newestTask = task("2023-01-04")
    const res = _test.getPriod([
      oldestTask,
      task("2023-01-02"),
      newestTask,
    ], date("2023-01-04"))

    expect(res).toEqual([
      date("2023-01-01"),
      date("2023-01-02"),
      date("2023-01-03"),
      date("2023-01-04")
    ])
  })
  test("getPriod - 2 ", () => {
    const oldestTask = task("2023-01-01")
    const newestTask = task("2023-01-04")
    const res = _test.getPriod([
      oldestTask,
      task(),
      task(),
      task(),
      task(),
    ], date("2023-01-02"))

    expect(res).toEqual([
      date("2023-01-01"),
      date("2023-01-02"),
    ])
  })

  test("getPriodDate - 4 day", () => {
    const res = _test.getPriodDate(
      date("2023-01-01"),
      date("2023-01-04"),
    )
    expect(res).toEqual([
      date("2023-01-01"),
      date("2023-01-02"),
      date("2023-01-03"),
      date("2023-01-04"),
    ])
  })
  test("getPriodDate - 1 day", () => {
    const res = _test.getPriodDate(
      date("2023-01-01"),
      date("2023-01-01"),
    )
    expect(res).toEqual([
      date("2023-01-01"),
    ])
  })

  test("getRemainTasksPerDay", () => {
    const res = _test.getRemainTasksPerDay(
      [
        task("2023-01-01"),
        task("2023-01-02"),
        task("2023-01-03"),
      ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
      ]
    )
    expect(res).toEqual([2, 1, 0])
  })

  test("getRemainTasksPerDay - 3 task", () => {
    const res = _test.getRemainTasksPerDay(
      [
        task("2023-01-01"),
        task(),
        task("2023-01-03"),
      ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
      ]
    )
    expect(res).toEqual([2, 2, 1])
  })

  test("getRemainTasksPerDay - 4 task", () => {
    const res = _test.getRemainTasksPerDay(
      [
        task("2023-01-02"),
        task(),
        task(),
        task("2023-01-03"),
      ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
        date("2023-01-04"),
      ]
    )
    expect(res).toEqual([4, 3, 2, 2])
  })
  test("getRemainTasksPerDay - 4 task", () => {
    const res = _test.getRemainTasksPerDay(
      [
        task("2023-01-01"),
        task("2023-01-02"),
        task("2023-01-03"),
        task("2023-01-04"),
      ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
        date("2023-01-04"),
      ]
    )
    expect(res).toEqual([3, 2, 1, 0])
  })
  test("getRemainTasksPerDay - 8 task", () => {
    const res = _test.getRemainTasksPerDay(
      [
        task("2023-01-01"),
        task("2023-01-02"),
        task("2023-01-03"),
        task(),
        task(),
        task(),
        task(),
        task(),
      ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
        date("2023-01-04"),
      ]
    )
    expect(res).toEqual([7, 6, 5, 5])
  })


  test("getPlandedTasksPerDay", () => {
    const res = _test.getPlandedTasksPerDay([
      task("2023-01-02"),
      task(),
      task("2023-01-03"),
    ],
      [
        date("2023-01-01"),
        date("2023-01-02"),
        date("2023-01-03"),
      ]
    )
    expect(res).toEqual([2, 1, 0])
  })

})
