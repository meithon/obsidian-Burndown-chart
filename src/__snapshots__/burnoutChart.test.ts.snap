// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`2 + 2 1`] = `
{
  "completedTasksPerDay": [],
  "labels": [],
  "plannedTasksPerDay": [],
}
`;
